syntax = "proto3";

package dlvision;

service Preprocess {
	rpc Resize (dataloc) returns (TaskReply) {}
	rpc Normalize (dataloc) returns (TaskReply) {}
	rpc SegPreprocess(prepOptions) returns (PrepReply) {}
}

service ObjectSegmentation {
	rpc RCNNTrain (segTrainOptions) returns (TrainReply) {}
}

message dataloc {
	string type = 1;	// 'local', 'azure blob', 'aws s3', 'smb'
	string name = 2;	// folder name or uri
}

message prepOptions {
	string name = 1;
	string locationType=2;
	string location=3;
	bool include_masks = 4;
	string train_dir=5;		// output directories and files
	string val_dir=6;
	string test_dir=7;
	string train_anno=8;
	string val_anno=9;
	string train_cap_anno=10;
	string val_cap_anno=11;
	string testdev_anno=12;
}

message segTrainOptions {
	string name = 1;
	string locationType=2;
	string location=3;
	bool include_masks = 4;
	//string train_dir=4;	// assume ./train, ./val, ./test
	//string val_dir=5;
	//string test_dir=6;
	string train_anno=5;
	string val_anno=6;
	string train_cap_anno=7;
	string val_cap_anno=8;
	string testdev_anno=9;
	int32 batch_size=10;
	bool no_xla=11;
	bool use_amp=12;
	int32 num_classes=13;
	ModelParams model_params = 14;
	int32 tb_port = 15;
}
// so many hyper-parameters... so let's pause for now

message TaskReply {
	int32 error = 1;
	string resultLoc = 2;
	repeated string msgs =3;
}

message PrepReply {
	int32 error = 1;
	string msg = 2;
	string location = 3;	// train folder is train, validation folder is val, test is test
}

message TrainReply {
	int32 error = 1;
	string msg = 2;
	string ckpt = 3;	// checkpoint folder
	string model_dir = 4;
}

message ModelParams {
	int32 min_level = 1;
	int32 max_level = 2;
	int32 num_scales = 3;
	int32 num_classes = 4;
	int32 fast_rcnn_mlp_head_dim = 5;
	int32 mrcnn_resolution = 6;
	int32 image_width=7;
	int32 image_height=8;
	bool skip_crowd=9;
	bool use_category=10;
	bool augment_input_data=11;
	//int32 gt_mast_size=12;
}